glotaran/testing/simple_generator.py

```diff
@@ -0,0 +1,207 @@
+    values: list = [val for val in input_list if not isinstance(val, dict)]
+    defaults: list = [val for val in input_list if isinstance(val, dict)]
+    return values, defaults
+
+
+class SimpleGeneratorError(Exception):
+    """Exception raised for errors in the simple_generator.
+
+    Attributes:
+        generator -- generator which is invalid
+        message -- explanation of the error
+    """
+
+    def __init__(self, message="Invalid generator state"):
+        self.message = message
+        super().__init__(self.message)
+
+
+@dataclass
+class SimpleGenerator:
+    """A minimal boilerplate model and parameters generator"""
+
+    rates: list[int | float] = field(default_factory=list)
+    """A list of values representing decay rates"""
+    k_matrix: Literal["parallel", "sequential"] | dict[tuple[str, str], str] = "parallel"
+    """"A `dict` with a k_matrix specification or `Literal["parallel", "sequential"]`"""
+    compartments: list[str] | None = None
+    irf: dict[str, float] = field(default_factory=dict)
+    initial_concentration: list[float] = field(default_factory=list)
+    dispersion_coefficients: list[float] = field(default_factory=list)
+    dispersion_center: float | None = None
+    default_megacomplex: str = "decay"
+    # TODO: add:
+    # shapes: list[float] = field(default_factory=list, init=False)
+
+    @property
+    def valid(self) -> bool:
+        try:
+            return self.model.valid(parameters=self.parameters)
+        except SimpleGeneratorError:
+            return False
+
+    def validate(self) -> str:
+        return self.model.validate(parameters=self.parameters)
+
+    @property
+    def model(self) -> Model:
+        return Model.from_dict(self.model_dict)
+
+    @property
+    def model_dict(self) -> dict:
+        return self._model_dict()
+
+    @property
+    def parameters(self) -> ParameterGroup:
+        return ParameterGroup.from_dict(self.parameters_dict)
+
+    @property
+    def parameters_dict(self) -> dict:
+        return self._parameters_dict()
+
+    @property
+    def model_and_parameters(self) -> tuple[Model, ParameterGroup]:
+        return self.model, self.parameters
+
+    def _validate_rates(self) -> tuple[bool, str]:
+        if not isinstance(self.rates, list):
+            return False, f"generator.rates: must be a `list`, got: {self.rates}"
+        if len(self.rates) == 0:
+            return False, "generator.rates: must be a `list` with 1 or more rates"
+        if not isinstance(self.rates[0], (int, float)):
+            return False, f"generator.rates: 1st element must be numeric, got: {self.rates[0]}"
+        return True, "generators.rates are valid"
+
+    @property
+    def _rates(self):
+        rates_valid, message = self._validate_rates()
+        if rates_valid:
+            return _split_iterable_in_values_and_dicts(self.rates)
+        else:
+            raise SimpleGeneratorError(message=message)
```

Comment:
Since `_validate_rates` only gets called in `_rates` we could combine it to `_valid_rates` and just raise at each validation step, that way the whole body of `_rates` isn't needed.

Reply:
Resolved in 63ee3818bcf7a07d9998dd7d1a5325e7c8ae944c